storage: TestStoreRangeWaitForApplication failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/0f5a289d49105cd317a2713b71546d84c25547f6

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeWaitForApplication PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=937734&tab=buildLog

```
=== RUN   TestStoreRangeWaitForApplication
I181002 05:30:59.913562 76254 gossip/gossip.go:393  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:36587" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
W181002 05:30:59.919745 76254 gossip/gossip.go:1496  [n2] no incoming or outgoing connections
I181002 05:30:59.919844 76254 gossip/gossip.go:393  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:43919" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I181002 05:30:59.920077 76359 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:36587
W181002 05:30:59.926382 76254 gossip/gossip.go:1496  [n3] no incoming or outgoing connections
I181002 05:30:59.926490 76254 gossip/gossip.go:393  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:41723" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I181002 05:30:59.926668 76362 gossip/client.go:129  [n3] started gossip client to 127.0.0.1:36587
I181002 05:30:59.982767 76254 storage/replica_command.go:298  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "a" [r2]
I181002 05:30:59.987155 76254 rpc/nodedialer/nodedialer.go:92  [s1,r2/1:{a-/Max}] connection to n2 established
I181002 05:30:59.987878 76254 storage/store_snapshot.go:615  [s1,r2/1:{a-/Max}] sending preemptive snapshot a9e313a6 at applied index 11
I181002 05:30:59.990504 76254 storage/store_snapshot.go:657  [s1,r2/1:{a-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 1, rate-limit: 2.0 MiB/sec, 3ms
I181002 05:30:59.991343 76481 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 11 (id=a9e313a6, encoded size=7467, 1 rocksdb batches, 1 log entries)
I181002 05:30:59.991646 76481 storage/replica_raftstorage.go:809  [s2,r2/?:{a-/Max}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I181002 05:30:59.992582 76254 storage/replica_command.go:812  [s1,r2/1:{a-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{a-/Max} [(n1,s1):1, next=2, gen=0]
I181002 05:30:59.994502 76254 storage/replica.go:3820  [s1,r2/1:{a-/Max},txn=6bd05900] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181002 05:30:59.995227 76254 rpc/nodedialer/nodedialer.go:92  [s1,r2/1:{a-/Max}] connection to n3 established
I181002 05:30:59.995638 76254 storage/store_snapshot.go:615  [s1,r2/1:{a-/Max}] sending preemptive snapshot 30e14ae5 at applied index 14
I181002 05:30:59.995897 76254 storage/store_snapshot.go:657  [s1,r2/1:{a-/Max}] streamed snapshot to (n3,s3):?: kv pairs: 44, log entries: 4, rate-limit: 2.0 MiB/sec, 1ms
I181002 05:30:59.996292 76611 storage/replica_raftstorage.go:803  [s3,r2/?:{-}] applying preemptive snapshot at index 14 (id=30e14ae5, encoded size=8360, 1 rocksdb batches, 4 log entries)
I181002 05:30:59.996884 76611 storage/replica_raftstorage.go:809  [s3,r2/?:{a-/Max}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I181002 05:30:59.997255 76254 storage/replica_command.go:812  [s1,r2/1:{a-/Max}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r2:{a-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I181002 05:31:00.000146 76394 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I181002 05:31:00.002431 76254 storage/replica.go:3820  [s1,r2/1:{a-/Max},txn=726e462c] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
--- FAIL: TestStoreRangeWaitForApplication (0.15s)
	client_raft_test.go:4069: 2: WaitForApplication returned early (request: 3, err: <nil>)

```

--- a/pkg/storage/client_raft_test.go
+++ b/pkg/storage/client_raft_test.go
@@ -3980,10 +3980,26 @@ func TestStoreRangeRemovalCompactionSuggestion(t *testing.T) {
 func TestStoreRangeWaitForApplication(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
+	var filterEnabled int32
+
 	ctx := context.Background()
 	sc := storage.TestStoreConfig(nil)
 	sc.TestingKnobs.DisableReplicateQueue = true
 	sc.TestingKnobs.DisableReplicaGCQueue = true
+	sc.TestingKnobs.TestingRequestFilter = func(ba roachpb.BatchRequest) *roachpb.Error {
+		if ba.RangeID != 2 || atomic.LoadInt32(&filterEnabled) == 0 {
+			return nil
+		}
+		pErr := roachpb.NewErrorf("blocking %s in this test", ba.Summary())
+		if len(ba.Requests) != 1 {
+			return pErr
+		}
+		_, ok := ba.Requests[0].GetInner().(*roachpb.PutRequest)
+		if !ok {
+			return pErr
+		}
+		return nil
+	}
 	mtc := &multiTestContext{storeConfig: &sc}
 	mtc.Start(t, 3)
 	defer mtc.Stop()
@@ -4003,6 +4019,9 @@ func TestStoreRangeWaitForApplication(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
+
+	atomic.StoreInt32(&filterEnabled, 1)
+
 	leaseIndex0 := repl0.LastAssignedLeaseIndex()
 
 	type target struct {
@@ -4081,6 +4100,8 @@ func TestStoreRangeWaitForApplication(t *testing.T) {
 			t.Fatalf("%d: %s", i, err)
 		}
 	}
+
+	atomic.StoreInt32(&filterEnabled, 0)
 
 	// GC the replica while a request is in progress. The request should return
 	// an error.

